lib/cloud/azure/client_map.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/gravitational/trace"
+)
+
+// ClientMap is a generic map that caches a collection of Azure clients by
+// subscriptions.
+type ClientMap[ClientType any] struct {
+	mu        sync.RWMutex
+	clients   map[string]ClientType
+	newClient func(string, azcore.TokenCredential, *arm.ClientOptions) (ClientType, error)
+}
+
+// NewClientMap creates a new ClientMap.
+func NewClientMap[ClientType any](newClient func(string, azcore.TokenCredential, *arm.ClientOptions) (ClientType, error)) ClientMap[ClientType] {
+	return ClientMap[ClientType]{
+		clients:   make(map[string]ClientType),
+		newClient: newClient,
+	}
+}
+
+// Get returns an Azure client by subscription. A new client is created if the
+// subscription is not found in the map.
+func (m *ClientMap[ClientType]) Get(subscription string, getCredentials func() (azcore.TokenCredential, error)) (client ClientType, err error) {
+	m.mu.RLock()
+	if client, ok := m.clients[subscription]; ok {
+		m.mu.RUnlock()
+		return client, nil
+	}
+	m.mu.RUnlock()
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
```

Comment:
Nit: Since the m.mux is read locked for single followed by RWLock we can optimise readability vs performance and squash lock to RWLock/RWUnlock calls: 
```suggestion
	m.mu.Lock()
	defer m.mu.Unlock()
	if client, ok := m.clients[subscription]; ok {
		return client, nil
	}
```

Reply:
Thanks for the suggestion. I think I will keep it the current way as the `RLock` allows multiple readers at the same time which is a possible scenario (e.g two redis get token at the same time).